File: scripts/generate_readme_from_template.js
const fs = require('fs')

function extractCurrentValue(readmeText, regex, fallback = 0) {
  const match = readmeText.match(regex)
  return match ? Number(match[1]) : fallback
}

const sleep = ms => new Promise(resolve => setTimeout(resolve, ms))

async function runRequest(github, headers, core, route, params, { attempts = 3, fallback = null } = {}) {
  let lastError

  for (let i = 1; i <= attempts; i++) {
    try {
      return await github.request(route, {
        ...params,
        headers: {
          ...headers,
          ...(params?.headers || {}),
        },
      })
    } catch (error) {
      lastError = error
      core.warning(`Request attempt ${i}/${attempts} failed for ${route}: ${error.message}`)
      if (i < attempts) {
        await sleep(1000 * i)
      }
    }
  }

  if (fallback !== null) {
    core.warning(`Falling back for ${route} due to failure: ${lastError?.message || 'unknown error'}`)
    return fallback
  }

  throw lastError
}

async function run({ github, core, context }) {
  const token = process.env.STATS_TOKEN
  if (!token) {
    core.setFailed('Missing token: define USER_TOKEN or allow GITHUB_TOKEN fallback')
    return
  }

  const headers = { authorization: `token ${token}` }
  const readmePath = `./${process.env.TARGET_FILE || 'README.md'}`
  const templatePath = './TEMPLATE.md'
  const readmeText = fs.existsSync(readmePath) ? fs.readFileSync(readmePath, 'utf8') : ''

  const current = {
    ISSUES: extractCurrentValue(readmeText, /I've opened\s+(\d+)\s+issues/i),
    PULL_REQUESTS: extractCurrentValue(readmeText, /I've contributed with\s+(\d+)\s+pull requests/i),
    COMMITS: extractCurrentValue(readmeText, /I've made\s+(\d+)\s+commits/i),
    REPOSITORIES_CONTRIBUTED_TO: extractCurrentValue(readmeText, /distributed amongst\s+(\d+)\s+repos/i),
  }

  const owner = context.repo.owner

  const issuesRes = await runRequest(
    github,
    headers,
    core,
    'GET /search/issues',
    { q: `author:${owner} is:issue`, per_page: 1 },
    { fallback: { data: { total_count: current.ISSUES } } }
  )

  const prsRes = await runRequest(
    github,
    headers,
    core,
    'GET /search/issues',
    { q: `author:${owner} is:pr`, per_page: 1 },
    { fallback: { data: { total_count: current.PULL_REQUESTS } } }
  )

  const commitsRes = await runRequest(
    github,
    headers,
    core,
    'GET /search/commits',
    {
      q: `author:${owner}`,
      per_page: 1,
      headers: { accept: 'application/vnd.github.cloak-preview+json' },
    },
    { fallback: { data: { total_count: current.COMMITS } } }
  )

  const replacements = {
    ISSUES: Number(issuesRes?.data?.total_count ?? current.ISSUES),
    PULL_REQUESTS: Number(prsRes?.data?.total_count ?? current.PULL_REQUESTS),
    COMMITS: Number(commitsRes?.data?.total_count ?? current.COMMITS),
    REPOSITORIES_CONTRIBUTED_TO: Number(current.REPOSITORIES_CONTRIBUTED_TO),
  }

  let template = fs.readFileSync(templatePath, 'utf8')
  template = template.replace(/\{\{\s*(ISSUES|PULL_REQUESTS|COMMITS|REPOSITORIES_CONTRIBUTED_TO)\s*\}\}/g, (_, key) => String(replacements[key]))
  fs.writeFileSync(readmePath, template)
}

module.exports = { run }

File: scripts/generate_readme_from_template.py
#!/usr/bin/env python3
from __future__ import annotations

import argparse
import json
import os
import re
import time
from pathlib import Path
from typing import Dict
from urllib import parse, request

ISSUES_RE = re.compile(r"I've opened\s+(\d+)\s+issues", re.IGNORECASE)
PRS_RE = re.compile(r"I've contributed with\s+(\d+)\s+pull requests", re.IGNORECASE)
COMMITS_RE = re.compile(r"I've made\s+(\d+)\s+commits", re.IGNORECASE)
CONTRIB_REPOS_RE = re.compile(r"distributed amongst\s+(\d+)\s+repos", re.IGNORECASE)


def extract_current_stats(text: str) -> Dict[str, int]:
    def get(regex: re.Pattern[str]) -> int:
        m = regex.search(text)
        return int(m.group(1)) if m else 0

    return {
        "ISSUES": get(ISSUES_RE),
        "PULL_REQUESTS": get(PRS_RE),
        "COMMITS": get(COMMITS_RE),
        "REPOSITORIES_CONTRIBUTED_TO": get(CONTRIB_REPOS_RE),
    }


def github_get(url: str, token: str) -> dict:
    req = request.Request(
        url,
        headers={
            "Accept": "application/vnd.github+json",
            "Authorization": f"Bearer {token}",
            "User-Agent": "scornejob-readme-test-generator",
            "X-GitHub-Api-Version": "2022-11-28",
        },
        method="GET",
    )
    with request.urlopen(req) as res:
        return json.loads(res.read().decode("utf-8"))


def with_retry(url: str, token: str, fallback: int) -> int:
    last_error = None
    for attempt in range(1, 4):
        try:
            data = github_get(url, token)
            return int(data.get("total_count", fallback))
        except Exception as error:  # noqa: BLE001
            last_error = error
            print(f"Request attempt {attempt}/3 failed for {url}: {error}")
            if attempt < 3:
                time.sleep(attempt)
    print(f"Falling back for {url}: {last_error}")
    return fallback


def main() -> int:
    parser = argparse.ArgumentParser(description="Generate README-like file from template")
    parser.add_argument("--template", default="TEMPLATE.md")
    parser.add_argument("--output", default="README-TEST.md")
    parser.add_argument("--token", default=os.environ.get("STATS_TOKEN") or os.environ.get("GITHUB_TOKEN", ""))
    args = parser.parse_args()

    template_text = Path(args.template).read_text(encoding="utf-8")
    existing_output_text = Path(args.output).read_text(encoding="utf-8") if Path(args.output).exists() else ""
    current = extract_current_stats(existing_output_text)

    owner = (os.environ.get("GITHUB_REPOSITORY") or "").split("/")[0]
    if not owner:
        raise RuntimeError("GITHUB_REPOSITORY is required")

    replacements = dict(current)
    if args.token:
        issues_q = parse.quote(f"author:{owner} is:issue", safe="")
        prs_q = parse.quote(f"author:{owner} is:pr", safe="")
        commits_q = parse.quote(f"author:{owner}", safe="")

        replacements["ISSUES"] = with_retry(
            f"https://api.github.com/search/issues?q={issues_q}&per_page=1",
            args.token,
            current["ISSUES"],
        )
        replacements["PULL_REQUESTS"] = with_retry(
            f"https://api.github.com/search/issues?q={prs_q}&per_page=1",
            args.token,
            current["PULL_REQUESTS"],
        )
        replacements["COMMITS"] = with_retry(
            f"https://api.github.com/search/commits?q={commits_q}&per_page=1",
            args.token,
            current["COMMITS"],
        )

    output = re.sub(
        r"\{\{\s*(ISSUES|PULL_REQUESTS|COMMITS|REPOSITORIES_CONTRIBUTED_TO)\s*\}\}",
        lambda m: str(replacements[m.group(1)]),
        template_text,
    )
    Path(args.output).write_text(output, encoding="utf-8")
    print(f"Generated {args.output} from {args.template}")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
